### ISSUE TITLE
AttributeError: 'TableRowGroupBox' object has no attribute 'index' / version 51 and 50

### ISSUE THREAD

**Issue by pulse-mind** [Kozea/WeasyPrint#1027]
Hello,
Thanks for your job. 

It works fine with WeasyPrint==49.

This is the error with version 50 and 51 : 
```
AttributeError: 'TableRowGroupBox' object has no attribute 'index'
  File "django/core/handlers/exception.py", line 41, in inner
    response = get_response(request)
  File "django/core/handlers/base.py", line 217, in _get_response
    response = self.process_exception_by_middleware(e, request)
  File "django/core/handlers/base.py", line 215, in _get_response
    response = response.render()
  File "django/template/response.py", line 107, in render
    self.content = self.rendered_content
  File "django_weasyprint/views.py", line 87, in rendered_content
    document = self.get_document()
  File "django_weasyprint/views.py", line 80, in get_document
    return html.render(self.get_css(base_url, url_fetcher))
  File "weasyprint/__init__.py", line 168, in render
    font_config)
  File "weasyprint/document.py", line 393, in _render
    [Page(page_box, enable_hinting) for page_box in page_boxes],
  File "weasyprint/document.py", line 393, in <listcomp>
    [Page(page_box, enable_hinting) for page_box in page_boxes],
  File "weasyprint/layout/__init__.py", line 126, in layout_document
    pages = list(make_all_pages(context, root_box, html, pages))
  File "weasyprint/layout/pages.py", line 804, in make_all_pages
    page, resume_at = remake_page(i, context, root_box, html)
  File "weasyprint/layout/pages.py", line 743, in remake_page
    page_number, page_state)
  File "weasyprint/layout/pages.py", line 554, in make_page
    positioned_boxes, adjoining_margins)
  File "weasyprint/layout/blocks.py", line 63, in block_level_layout
    page_is_empty, absolute_boxes, fixed_boxes, adjoining_margins)
  File "weasyprint/layout/blocks.py", line 77, in block_level_layout_switch
    page_is_empty, absolute_boxes, fixed_boxes, adjoining_margins)
  File "weasyprint/layout/blocks.py", line 130, in block_box_layout
    absolute_boxes, fixed_boxes, adjoining_margins)
  File "weasyprint/layout/blocks.py", line 507, in block_container_layout
    absolute_boxes, fixed_boxes, adjoining_margins)
  File "weasyprint/layout/blocks.py", line 63, in block_level_layout
    page_is_empty, absolute_boxes, fixed_boxes, adjoining_margins)
  File "weasyprint/layout/blocks.py", line 77, in block_level_layout_switch
    page_is_empty, absolute_boxes, fixed_boxes, adjoining_margins)
  File "weasyprint/layout/blocks.py", line 130, in block_box_layout
    absolute_boxes, fixed_boxes, adjoining_margins)
  File "weasyprint/layout/blocks.py", line 552, in block_container_layout
    new_children, absolute_boxes, fixed_boxes)
  File "weasyprint/layout/blocks.py", line 818, in find_earlier_page_break
    child.children, absolute_boxes, fixed_boxes)
  File "weasyprint/layout/blocks.py", line 818, in find_earlier_page_break
    child.children, absolute_boxes, fixed_boxes)
  File "weasyprint/layout/blocks.py", line 809, in find_earlier_page_break
    resume_at = (children[index].index, None)
```

And this is the breadcrumbs: 
```
django.template Exception while resolving variable 'additional' in template 'admin/filter.html'.
django.template Exception while resolving variable 'date_range_filter' in template 'admin/softdeletable_change_list_import_export.html'.
django.template Exception while resolving variable 'errors' in template 'admin/softdeletable_change_list_import_export.html'.
django.template Exception while resolving variable 'is_multipart' in template 'admin/softdeletable_change_list_import_export.html'.
django.template Exception while resolving variable 'non_field_errors' in template 'admin/softdeletable_change_list_import_export.html'.
django.template Exception while resolving variable 'separator' in template 'admin/softdeletable_change_list_import_export.html'.
sqlqueries Request Time: 1.021 seconds
gunicorn.error GET /app/documents/xxx/xxx/pdf/
query SELECT typarray FROM pg_type WHERE typname = 'citext'
query [Filtered]
PDFView : model_name : mymodel
django.template exception while resolving variable 'title' in template 'documents/doc_pdf.html'.
django.template Exception while resolving variable 'showsidebar' in template 'documents/doc_pdf.html'.
django.template Exception while resolving variable 'enable_eco_tax' in template 'documents/doc_pdf.html'.
weasyprint.progress Step 1 - Fetching and parsing HTML - HTML string
weasyprint.progress Step 2 - Fetching and parsing CSS - CSS string
weasyprint Ignored `-moz-column-count: 2` at 702:7, prefixed selectors are ignored.
weasyprint Ignored `-moz-column-gap: 10px` at 703:7, prefixed selectors are ignored.
weasyprint Ignored `-webkit-column-count: 2` at 704:7, prefixed selectors are ignored.
weasyprint Ignored `-webkit-column-gap: 10px` at 705:7, prefixed selectors are ignored.
weasyprint.progress Step 2 - Fetching and parsing CSS - CSS string
weasyprint.progress Step 3 - Applying CSS
weasyprint.progress Step 4 - Creating formatting structure
weasyprint.progress Step 5 - Creating layout - Page 1
exception AttributeError: 'TableRowGroupBox' object has no attribute 'index'
```
On Django 1.11.27 and django-weasyprint==0.5.5

**Comment by vog** [comment]
I experience the same problem, but can't easily revert to WeasyPrint 49, as that version doesn't provide some page (non-)breaking mechanisms that is important for me.

Has there been any progress on this?

Do you have any hints for where to dig into? Or, is the stack trace sufficient to find all relevant places in the code that need to be fixed?

**Comment by fbierman** [comment]
Hello,
The context in which this issue occurs for me is as such:
- There are 2 blocks, let's call them block1 and block2
- Block1 last element is a table **with an empty footer** -> the issue does not occur if the table doesn't have a footer
- There is a non-page-breaking rule between block1 and block2
- Block1 is printed until the end of page x and block2 is printed at the beginning of page x+1

Here's an example of a very basic HTML file triggering this bug:
```
<!DOCTYPE html>
<html>
<head>
<style>
table {
  border: 0.2mm solid;
  table-layout: fixed;
  font-family: arial, sans-serif;
  border-collapse: collapse;
  width: 100%;
}

td {
  text-align: left;
  padding: 8px;
}

.avoid-break-before {
  break-before: avoid;
}
</style>
</head>
<body>
  <table class="avoid-break-before">
    <tbody>
      <tr>
        <td>value<div>1</div><div>1</div></td>
        <td>value<div>1</div><div>2</div></td>
        <td>value<div>1</div><div>3</div></td>
      </tr>
      <tr>
        <td>value<div>2</div><div>1</div></td>
        <td>value<div>2</div><div>2</div></td>
        <td>value<div>2</div><div>3</div></td>
      </tr>
      <tr>
        <td>value<div>3</div><div>1</div></td>
        <td>value<div>3</div><div>2</div></td>
        <td>value<div>3</div><div>3</div></td>
      </tr>
      <tr>
        <td>value<div>4</div><div>1</div></td>
        <td>value<div>4</div><div>2</div></td>
        <td>value<div>4</div><div>3</div></td>
      </tr>
      <tr>
        <td>value<div>5</div><div>1</div></td>
        <td>value<div>5</div><div>2</div></td>
        <td>value<div>5</div><div>3</div></td>
      </tr>
    </tbody>
    <tfoot>
    </tfoot>
   </table>
  <table class="avoid-break-before">
    <tbody>
      <tr>
        <td>value<div>1</div><div>1</div></td>
        <td>value<div>1</div><div>2</div></td>
        <td>value<div>1</div><div>3</div></td>
      </tr>
      <tr>
        <td>value<div>2</div><div>1</div></td>
        <td>value<div>2</div><div>2</div></td>
        <td>value<div>2</div><div>3</div></td>
      </tr>
      <tr>
        <td>value<div>3</div><div>1</div></td>
        <td>value<div>3</div><div>2</div></td>
        <td>value<div>3</div><div>3</div></td>
      </tr>
      <tr>
        <td>value<div>4</div><div>1</div></td>
        <td>value<div>4</div><div>2</div></td>
        <td>value<div>4</div><div>3</div></td>
      </tr>
      <tr>
        <td>value<div>5</div><div>1</div></td>
        <td>value<div>5</div><div>2</div></td>
        <td>value<div>5</div><div>3</div></td>
      </tr>
      <tr>
        <td>value<div>6</div><div>1</div></td>
        <td>value<div>6</div><div>2</div></td>
        <td>value<div>6</div><div>3</div></td>
      </tr>
      <tr>
        <td>value<div>7</div><div>1</div></td>
        <td>value<div>7</div><div>2</div></td>
        <td>value<div>7</div><div>3</div></td>
      </tr>
      <tr>
        <td>value<div>8</div><div>1</div></td>
        <td>value<div>8</div><div>2</div></td>
        <td>value<div>8</div><div>3</div></td>
      </tr>
    </tbody>
    <tfoot>
    </tfoot>
   </table>
  <table class="avoid-break-before">
    <tbody>
      <tr>
        <td>value<div>1</div><div>1</div></td>
        <td>value<div>1</div><div>2</div></td>
        <td>value<div>1</div><div>3</div></td>
      </tr>
      <tr>
        <td>value<div>2</div><div>1</div></td>
        <td>value<div>2</div><div>2</div></td>
        <td>value<div>2</div><div>3</div></td>
      </tr>
      <tr>
        <td>value<div>3</div><div>1</div></td>
        <td>value<div>3</div><div>2</div></td>
        <td>value<div>3</div><div>3</div></td>
      </tr>
    </tbody>
    <tfoot>
    </tfoot>
   </table>
</body>
</html>
```
The first two tables are printed on the first page while the third table has to be printed on the second page. However, the rule `break-before: avoid` prevents the third table to be printed alone on the page, provoking the bug.

When looking into the code, this issue occurs in the function `find_earlier_page_break` located in the `blocks.py` module. While searching for a place to break in the previous table, its footer has no index for some reason. Providing a default index to the footer was enough to make the code run but the behavior is not as expected from my point of view because the non-page-breaking rule is ignored.
For me, the expected behavior would be:
- The last row of the second table goes on the next page with the third table
- The footer of the second table is placed after the penultimate row (on page x) and after the last row (on page x+1) of the second table

I wrote a very specific workaround but I hope this code can be used to fix this issue. Here's the code of the `find_earlier_page_break` function I modified:
```
def find_earlier_page_break(children, absolute_boxes, fixed_boxes):
    if children and isinstance(children[0], boxes.LineBox):
        # Normally `orphans` and `widows` apply to the block container, but
        # line boxes inherit them.
        orphans = children[0].style['orphans']
        widows = children[0].style['widows']
        index = len(children) - widows  # how many lines we keep
        if index < orphans:
            return None
        new_children = children[:index]
        resume_at = (0, new_children[-1].resume_at)
        remove_placeholders(children[index:], absolute_boxes, fixed_boxes)
        return new_children, resume_at

    previous_in_flow = None
    for index, child in reversed_enumerate(children):
        if child.is_in_normal_flow():
            if previous_in_flow is not None and (
                    block_level_page_break(child, previous_in_flow) not in
                    ('avoid', 'avoid-page')):
                # New code
                if isinstance(child, boxes.TableRowGroupBox):
                    result = find_earlier_page_break(
                        child.children, absolute_boxes, fixed_boxes)
                    if result:
                        new_grand_children, resume_at = result
                        new_child = child.copy_with_children(new_grand_children)
                        old_height = new_child.height
                        new_height = child.children[resume_at[0] - child.children[0].index].position_y - new_child.position_y
                        new_child.height = new_height
                        diff_height = old_height - new_height
                        for child in children[index+1:]:
                            for descendant in child.descendants():
                                descendant.position_y -= diff_height
                        new_children = list(children[:index]) + [new_child] + list(children[index+1:])
                        resume_at = (new_child.index, resume_at)
                        index += 1  # Remove placeholders after child
                        break
                else:
                # End of new code
                    index += 1  # break after child
                    new_children = children[:index]
                    # Get the index in the original parent
                    resume_at = (children[index].index, None)
                    break
            previous_in_flow = child
        if child.is_in_normal_flow() and (
                child.style['break_inside'] not in ('avoid', 'avoid-page')):
            breakable_box_types = (
                boxes.BlockBox, boxes.TableBox, boxes.TableRowGroupBox)
            if isinstance(child, breakable_box_types):
                result = find_earlier_page_break(
                    child.children, absolute_boxes, fixed_boxes)
                if result:
                    new_grand_children, resume_at = result
                    new_child = child.copy_with_children(new_grand_children)
                    new_children = list(children[:index]) + [new_child]
                    # Index in the original parent
                    resume_at = (new_child.index, resume_at)
                    index += 1  # Remove placeholders after child
                    break
    else:
        return None

    remove_placeholders(children[index:], absolute_boxes, fixed_boxes)
    return new_children, resume_at
```
The search used to stop at the body of the table which was causing the issue (+unexpected behavior if the footer has an index). Adding a recursive call allows to continue the search in the body for the last breakable row of the table. When this row is found then the footer is appended to the list of new children to satisfy the second requirement of the expected behavior. But the position of the footer is not updated automatically, so I had to change manually all the positions of the footer's sub-boxes.

I hope this can help !

**Comment by fbierman** [comment]
Here's patch preventing the crash to happen:
[prevent_crash.patch.gz](https://github.com/Kozea/WeasyPrint/files/4403947/prevent_crash.patch.gz)

**Comment by liZe** [comment]
Thanks a lot for your example, that’s a pleasure to work with both the code and your comment :smile:.

> For me, the expected behavior would be:
> 
> * The last row of the second table goes on the next page with the third table
> * The footer of the second table is placed after the penultimate row (on page x) and after the last row (on page x+1) of the second table

I agree, that’s how it should work.

> I hope this can help !

It definitely does.

I think that the problem happens because the layout of the footer has not been done "normally", as we didn’t reach the end of the table yet. As the footer is always drawn, it’s in the table children, but its index has not been set.

When we try to find a place where it’s possible to break a table, we should actually avoid table headers and footers, as we never want to break the table after the header or before the footer.

**Comment by liZe** [comment]
I’ve fixed your problem in da146c6, but it’s not perfect.

This fix removes the footer if the table can be drawn on the first page but has to be split because you avoid breaks between tables. Your example work, but the test fails. I have to find more time to fix this bug, I’ll be back soon :smile:.

**Comment by liZe** [comment]
WeasyPrint shouldn’t crash anymore, and you should have right data on each page. Some use cases should work perfectly now.

Unfortunately, footers may have wrong positions in split tables. That’s a bit hard to fix, and we should also have a problem with parent boxes with cloned decorations. If anyone has the problem, feel free to open a new bug with HTML/CSS samples.
